client/utils/src/notification.rs

```diff
@@ -0,0 +1,172 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provides mpsc notification channel that can be instantiated
+//! _after_ it's been shared to the consumer and producers entities.
+//!
+//! Useful when building RPC extensions where, at service definition time, we
+//! don't know whether the specific interface where the RPC extension will be
+//! exposed is safe or not and we want to lazily build the RPC extension
+//! whenever we bind the service to an interface.
+//!
+//! See [`sc-service::builder::RpcExtensionBuilder`] for more details.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use crate::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+
+use parking_lot::Mutex;
+
+/// Collection of channel sending endpoints shared with the receiver side
+/// so they can register themselves.
+type SharedSenders<Payload> = Arc<Mutex<Vec<TracingUnboundedSender<Payload>>>>;
+
+/// Trait used to define the "tracing key" string used to tag
+/// and identify the mpsc channels;
+pub trait TracingKeyStr {
+	const TRACING_KEY: &'static str;
+}
+
+/// The sending half of the notifications channel(s).
+///
+/// Used to send notifications from the BEEFY gadget side.
+pub struct NotificationSender<Payload: Clone, Error> {
+	subscribers: SharedSenders<Payload>,
+	_err: PhantomData<Error>,
+}
+
+// Because of https://github.com/rust-lang/rust/issues/26925 we need to
+// manually implement `clone` here.
+impl<Payload: Clone, Error> Clone for NotificationSender<Payload, Error> {
+	fn clone(&self) -> Self {
+		Self { subscribers: self.subscribers.clone(), _err: PhantomData }
+	}
+}
+
+impl<Payload: Clone, Error> NotificationSender<Payload, Error> {
+	/// The `subscribers` should be shared with a corresponding `NotificationStream`.
+	fn new(subscribers: SharedSenders<Payload>) -> Self {
+		Self { subscribers, _err: PhantomData }
+	}
+
+	/// Send out a notification to all subscribers that a new payload is available for a
+	/// block.
+	pub fn notify(&self, payload: impl FnOnce() -> Result<Payload, Error>) -> Result<(), Error> {
+		let mut subscribers = self.subscribers.lock();
+
+		// do an initial prune on closed subscriptions
+		subscribers.retain(|n| !n.is_closed());
```

Comment:
Alas, I didn't add a generic utility to clean up the subscriptions when the receiver is dropped, although that *might* actually be a good idea seeing that we have a very similar situation here.

What I did is basically [I wrapped the `TracingUnboundedReceiver` into a newtype and made it unsubscribe on drop](https://github.com/paritytech/substrate/pull/10454/files#diff-ece03d48ef90111c19d13e2b3a77068eeda4eaedad0b033dd9b18303bf63b538R84).

So I see four possibilities here:

1) Ignore the problem. *If* the notifications are sent relatively often then it won't become a critical problem. (Not a fan of this one though, as it just creates a subtle footgun which can blow up someday.)
2) Create another single-use newtype to use here.
3) Create a generic newtype to use both here, and in storage subscribers where I've previously fixed this problem.
4) Refactor the sender/receiver types (`TracingUnboundReceiver` etc.) so that they have this functionality built-in, so that everywhere they're used we get an unsubscribe-on-drop behavior essentially for free.

Personally I'd probably go for (4), although that would be the most work.

Reply:
Opened https://github.com/paritytech/substrate/issues/10576 and will address in a following PR.